akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -116,7 +117,39 @@ private[akka] trait InboundContext {
  */
 private[akka] object AssociationState {
   def apply(): AssociationState =
-    new AssociationState(incarnation = 1, uniqueRemoteAddressPromise = Promise(), quarantined = Set.empty)
+    new AssociationState(incarnation = 1, uniqueRemoteAddressPromise = Promise(), quarantined = QuarantinedUidSet.empty)
+
+  object QuarantinedUidSet {
+    val maxEntries = 10 // ok to not keep all old uids
+    def empty: QuarantinedUidSet = new QuarantinedUidSet(Array.emptyLongArray)
+  }
+
+  class QuarantinedUidSet private (uids: Array[Long]) {
+    import QuarantinedUidSet._
+
+    def add(uid: Long): QuarantinedUidSet = {
+      if (apply(uid))
+        this
+      else {
+        val newUids = Array.ofDim[Long](math.min(uids.length + 1, maxEntries))
+        newUids(0) = uid
+        if (uids.length > 0)
+          System.arraycopy(uids, 0, newUids, 1, newUids.length - 1)
+        new QuarantinedUidSet(newUids)
+      }
+    }
+
+    def apply(uid: Long): Boolean = {
```

Comment:
Alternative is to sort the array and do a binary search here (Arrays have both these methods).